src/combinatorics/simple_graph/regularity/uniform.lean

```diff
@@ -80,11 +85,90 @@ begin
   exact zero_lt_one,
 end
 
+variables {G}
+
+lemma exists_nonuniform_witnesses (h : ¬G.is_uniform ε s t) :
+  ∃ s', s' ⊆ s ∧ ∃ t', t' ⊆ t ∧ ↑s.card * ε ≤ s'.card ∧ ↑t.card * ε ≤ t'.card ∧
+    ε ≤ |G.edge_density s' t' - G.edge_density s t| :=
+by { rw is_uniform at h, push_neg at h, exact h }
+
+open_locale classical
+variables (G)
+
+/-- A pair of subsets witnessing the non-uniformity of `(s, t)`. If `(s, t)` is uniform, returns
+`(s, t)`. Witnesses for `(s, t)` and `(t, s)` don't necessarily match, hence the motivation to
+define `nonuniform_witness`. -/
+noncomputable def nonuniform_witnesses (ε : 𝕜) (s t : finset α) : finset α × finset α :=
+if h : ¬G.is_uniform ε s t
+  then ((exists_nonuniform_witnesses h).some, (exists_nonuniform_witnesses h).some_spec.2.some)
+  else (s, t)
+
+lemma left_nonuniform_witnesses_subset (h : ¬G.is_uniform ε s t) :
+  (G.nonuniform_witnesses ε s t).1 ⊆ s :=
+by { rw [nonuniform_witnesses, dif_pos h], exact (exists_nonuniform_witnesses h).some_spec.1 }
+
+lemma left_nonuniform_witnesses_card (h : ¬ G.is_uniform ε s t) :
+  (s.card : 𝕜) * ε ≤ (G.nonuniform_witnesses ε s t).1.card :=
+by { rw [nonuniform_witnesses, dif_pos h],
+  exact (exists_nonuniform_witnesses h).some_spec.2.some_spec.2.1 }
+
+lemma right_nonuniform_witnesses_subset (h : ¬ G.is_uniform ε s t) :
+  (G.nonuniform_witnesses ε s t).2 ⊆ t :=
+by { rw [nonuniform_witnesses, dif_pos h],
+  exact (exists_nonuniform_witnesses h).some_spec.2.some_spec.1 }
+
+lemma right_nonuniform_witnesses_card (h : ¬ G.is_uniform ε s t) :
+  (t.card : 𝕜) * ε ≤ (G.nonuniform_witnesses ε s t).2.card :=
+by { rw [nonuniform_witnesses, dif_pos h],
+  exact (exists_nonuniform_witnesses h).some_spec.2.some_spec.2.2.1 }
+
+lemma nonuniform_witnesses_spec (h : ¬ G.is_uniform ε s t) :
+  ε ≤ |G.edge_density (G.nonuniform_witnesses ε s t).1 (G.nonuniform_witnesses ε s t).2
+    - G.edge_density s t| :=
+by { rw [nonuniform_witnesses, dif_pos h],
+  exact (exists_nonuniform_witnesses h).some_spec.2.some_spec.2.2.2 }
+
+/-- Witnessing of non-uniformity of `(s, t)`. `G.nonuniform_witness ε s t` and
+`G.nonuniform_witness ε t s` form a pair of subsets witnessing the non-uniformity of `(s, t)`. If
+`(s, t)` is uniform, returns `s`. -/
+noncomputable def nonuniform_witness (ε : 𝕜) (s t : finset α) : finset α :=
+ite (well_ordering_rel s t) (G.nonuniform_witnesses ε s t).1 (G.nonuniform_witnesses ε t s).2
```

Comment:
I feel like mathlib prefers `if/then/else` over `ite`, but whatever the case that's at least my own preference due to readability :shrug:
```suggestion
if well_ordering_rel s t then (G.nonuniform_witnesses ε s t).1 else (G.nonuniform_witnesses ε t s).2
```